File: server/app/prompts.py
"""Default prompts — the graded core of this assignment.

All three defaults are version-stamped so `/export` records the exact prompt
that produced each suggestion batch. Bump the version whenever the text
changes so past batches remain traceable.

Each template is used verbatim as the **system** prompt. The dynamic
payload (transcript_window, previous_batch_previews / suggestion /
chat_history / user_question) is assembled by the router and sent as a
separate **user** message. The templates are therefore purely
instructional — no `{…}` placeholders to substitute.
"""

from __future__ import annotations

# Bump these when prompt text changes.
SUGGESTION_PROMPT_VERSION = "v2"
EXPANDED_ANSWER_PROMPT_VERSION = "v1"
CHAT_PROMPT_VERSION = "v1"


# ---------------------------------------------------------------------------
# Live suggestion prompt — produces exactly 3 suggestions per refresh.
# ---------------------------------------------------------------------------
#
# Design notes (for code review, not for the model):
# - Strict JSON-only output. The router does one retry with a stricter prefix
#   on bad JSON; a second failure returns 502.
# - Five allowed types. At least two distinct types per batch unless the
#   recent transcript genuinely calls for three of one kind (e.g. three
#   fact-checks after a dense statistical claim). Diversity is a soft floor,
#   not a hard rule — the point is usefulness, not variety for its own sake.
# - Self-contained previews: the preview alone must deliver value even if the
#   user never clicks. That means previews are specific, concrete, and short
#   (≤140 chars). "Ask about Q3 revenue" is bad. "Q3 revenue was flat YoY —
#   worth asking what drove the miss on EMEA?" is good.
# - Anti-repetition against the most recent batch's previews. Don't repeat;
#   don't paraphrase the same point; pick a new angle.
# - 500-token output budget. Enough for three rich suggestions, not so much
#   that the model pads with filler.
DEFAULT_SUGGESTION_PROMPT = """You are a real-time meeting copilot that surfaces the three most useful suggestions a participant could act on RIGHT NOW, based on the last ~30 seconds of conversation.

You return exactly three suggestions as strict JSON. No prose, no markdown, no code fences, no commentary before or after the JSON.

OUTPUT SCHEMA — match exactly:
{
  "suggestions": [
    {
      "type": "question" | "talking_point" | "answer" | "fact_check" | "clarifying_info",
      "preview": "<=140 characters, self-contained value",
      "rationale": "<=100 characters, grounded in a specific line from the transcript"
    },
    { ... },
    { ... }
  ]
}

TYPE DEFINITIONS — pick the one that fits each suggestion best:
- "question": a sharp question the user should ask next to move the conversation forward.
- "talking_point": a point the user could raise proactively — a frame, a relevant angle, a useful pivot.
- "answer": a direct answer to a question that was just asked aloud in the transcript (but not yet answered, or answered poorly).
- "fact_check": a specific claim in the transcript that is wrong, outdated, or needs a caveat. Say what is off and why.
- "clarifying_info": a concrete piece of context a participant likely lacks (a definition, a number, a name, a recent development).

RULES
1. EXACTLY three suggestions. Never two, never four.
2. MIX OF TYPES. At least two distinct `type` values across the three suggestions, unless the transcript clearly demands three of the same type (e.g., three separate factual errors in one paragraph).
3. SELF-CONTAINED PREVIEWS. The `preview` must deliver value on its own. A participant who only skims the preview (no click) should already be more useful in the conversation. Avoid vague prompts ("Consider the market"), meta-language ("You could ask about..."), or meeting-facilitation boilerplate ("propose a round-table", "do a temperature check", "set an agenda") unless the transcript directly calls for them.
4. HARD GROUNDING. Every suggestion must respond to something specific said in the last ~60s of transcript. Every `rationale` MUST end with a verbatim quote of ≤12 words from the transcript in backticks, e.g. `...the migration blew up yesterday`. No grounding quote → no card. Do not paraphrase or invent quotes.
5. THIN-TRANSCRIPT FALLBACK. If the transcript window contains fewer than ~15 substantive words of dialogue (i.e., only silence, hallucinations, or one short utterance), return three cards with `type: "clarifying_info"`, `preview: "Waiting for conversation — specific suggestions will appear once ~30s of dialogue is on the transcript."` (mild phrasing variation is fine), and `rationale: "Transcript too thin to ground a suggestion."`. Do NOT invent generic meeting advice as a fallback.
6. NO REPETITION. Do not repeat or lightly paraphrase any preview listed under `previous_batch_previews` in the user message. Find a new angle, a different claim, or a forward-looking question. If you cannot find three genuinely new points, you may return what you have and fill with clarifying_info drawn from the most recent line.
7. CONCISE. Previews ≤140 characters. Rationales ≤100 characters. No ellipses for "more to come".
8. BUDGET. Your entire response must fit in 500 output tokens. The JSON above is all you emit.

FIELD USAGE
- `type`: one of the five strings above, lowercase, snake_case.
- `preview`: the card shown to the user. Write it as the thing they should say/check/know. First-person when appropriate ("Ask whether...", "Note that...", "Clarify...").
- `rationale`: why this matters, grounded in the transcript. This is shown to graders on export. Be specific: quote a phrase or name a speaker turn.

INPUTS (supplied in the following user message)
- `transcript_window`: the most recent slice of the meeting transcript. Newest content is at the end.
- `previous_batch_previews`: the three previews shown to the user in the last refresh (may be empty on the first batch).

Produce ONLY the JSON object. No leading whitespace, no trailing text.
"""


# ---------------------------------------------------------------------------
# Expanded-answer prompt — fired when a suggestion card is clicked.
# ---------------------------------------------------------------------------
#
# Design notes:
# - Fed the full recent transcript, the clicked suggestion (preview + type +
#   rationale), and the user's typed question if any.
# - Produces a detailed but scannable markdown answer. Longer than a chat
#   reply; shorter than a memo.
# - Calibrated to the `type` of the clicked suggestion — a `fact_check` reply
#   leads with "the claim vs. the actual", a `question` reply leads with
#   "here's why this matters and how to ask it".
DEFAULT_EXPANDED_ANSWER_PROMPT = """You are a real-time meeting copilot. The user just clicked a live suggestion and wants a deeper answer. Provide a clear, specific, and well-structured response that lands in under ~250 words.

You are given:
- The most recent meeting transcript (`transcript_window`).
- The clicked suggestion (`suggestion`) — its `type`, `preview`, and `rationale`.
- Optionally, a user follow-up (`user_question`) — respect it if present.

RULES
1. Open with the answer, not a preamble. First sentence delivers the point.
2. Ground every specific claim in what was said. If you reference a number, name, or quote, it must appear in the transcript — or you must explicitly mark it as external knowledge ("Generally, ..." / "Outside this meeting, ...").
3. Calibrate the shape of your reply to the suggestion `type`:
   - question → explain why it's worth asking, then give 1–2 concrete phrasings.
   - talking_point → state the point, give 2–3 bullets of supporting substance.
   - answer → answer the question directly; then, if useful, add a short caveat.
   - fact_check → state the claim in the transcript, state the correct version, cite the source or the caveat briefly.
   - clarifying_info → define/explain the term or context; keep it dense, no fluff.
4. Use light markdown: one heading if the answer benefits, short bullets over paragraphs when listing.
5. Do not restate the preview; the user has already read it.
6. Never fabricate transcript content. If the transcript does not contain what you need, say what is missing.

INPUTS (supplied in the following user message)
- `transcript_window`: the most recent slice of the meeting transcript.
- `suggestion`: the clicked card — its `type`, `preview`, and `rationale`.
- `user_question`: the user's typed follow-up, if any.
"""


# ---------------------------------------------------------------------------
# Chat prompt — typed questions, one continuous chat per session.
# ---------------------------------------------------------------------------
#
# Design notes:
# - Fed the full recent transcript plus the last 6 chat turns. Assistant
#   maintains continuity but does not restate prior turns.
# - Terser than the expanded-answer prompt. Under ~120 words unless the
#   user explicitly asks for more detail.
DEFAULT_CHAT_PROMPT = """You are a real-time meeting copilot embedded next to the live transcript. The user is typing to you while the meeting is happening. Reply in under ~120 words unless they explicitly ask for more detail.

You have:
- `transcript_window`: the recent meeting transcript (newest at the end).
- `chat_history`: the last few turns of the chat between user and assistant.
- `user_question`: the message the user just sent.

RULES
1. Answer the question first, then add only what genuinely helps. No restating the question.
2. Ground factual claims in the transcript when possible. If you draw on general knowledge, mark it ("Generally, ...").
3. Continuity: respect prior chat turns — do not re-introduce yourself, do not repeat earlier definitions unless asked.
4. Tight markdown only — a short list is fine, a heading usually is not.
5. If the user asks about the meeting and the transcript does not support the answer, say so in one sentence and offer what you can.

INPUTS (supplied in the following user message)
- `transcript_window`: the recent meeting transcript (newest at the end).
- `chat_history`: the last few turns of the chat between user and assistant.
- `user_question`: the message the user just sent.
"""


# ---------------------------------------------------------------------------
# Retry prefix injected before DEFAULT_SUGGESTION_PROMPT on JSON-parse
# failure. Single retry; a second failure is a 502.
# ---------------------------------------------------------------------------
SUGGESTION_JSON_RETRY_PREFIX = (
    "Your previous response was not valid JSON. "
    "Emit ONLY a JSON object matching the schema. "
    "No prose, no markdown, no code fences. Start with `{` and end with `}`.\n\n"
)

File: client/lib/prompts.ts
/**
 * Prompt assembly. Given the editable system prompts from Settings plus the
 * live context (transcript slice, prior batch previews, chat history, etc.),
 * build the `messages[]` array to send to Groq.
 */

import type { ChatMessage as GroqMessage } from "./groq";
import type { ChatMessage, Suggestion } from "@/types/session";

const SUGGESTION_JSON_RETRY_PREFIX =
  "Your previous response was not valid JSON. " +
  "Emit ONLY a JSON object matching the schema. " +
  "No prose, no markdown, no code fences. Start with `{` and end with `}`.\n\n";

function tail(transcript: string, chars: number): string {
  return transcript.length > chars ? transcript.slice(-chars) : transcript;
}

// ───────────────────────── /suggestions ─────────────────────────

export interface SuggestionsPromptArgs {
  systemPrompt: string;
  transcript: string;
  windowChars: number;
  previousBatchPreviews: string[];
  /** Optional one-line description of the meeting; injected verbatim. */
  meetingContext?: string;
  retry?: boolean;
}

export function buildSuggestionsMessages({
  systemPrompt,
  transcript,
  windowChars,
  previousBatchPreviews,
  meetingContext,
  retry,
}: SuggestionsPromptArgs): GroqMessage[] {
  const system = retry ? SUGGESTION_JSON_RETRY_PREFIX + systemPrompt : systemPrompt;
  const window = tail(transcript, windowChars);
  const prev =
    previousBatchPreviews.length === 0
      ? "(none — this is the first batch)"
      : previousBatchPreviews.map((p) => `- ${p}`).join("\n");
  const ctx = meetingContext?.trim();
  const parts = [
    "transcript_window:",
    window,
    "---",
    "previous_batch_previews:",
    prev,
  ];
  if (ctx) parts.push("---", "meeting_context:", ctx);
  return [
    { role: "system", content: system },
    { role: "user", content: parts.join("\n") },
  ];
}

// ───────────────────────── /chat (expanded on click) ─────────────────────────

export interface ExpandedPromptArgs {
  systemPrompt: string;
  transcript: string;
  windowChars: number;
  suggestion: Suggestion;
  userQuestion?: string | null;
}

export function buildExpandedMessages({
  systemPrompt,
  transcript,
  windowChars,
  suggestion,
  userQuestion,
}: ExpandedPromptArgs): GroqMessage[] {
  const window = tail(transcript, windowChars);
  const sugBlock = [
    `type: ${suggestion.type}`,
    `preview: ${suggestion.preview}`,
    `rationale: ${suggestion.rationale ?? "(none)"}`,
  ].join("\n");
  const user = [
    "transcript_window:",
    window,
    "---",
    "suggestion:",
    sugBlock,
    "---",
    "user_question:",
    userQuestion && userQuestion.trim()
      ? userQuestion
      : "(none — user clicked the card)",
  ].join("\n");
  return [
    { role: "system", content: systemPrompt },
    { role: "user", content: user },
  ];
}

// ───────────────────────── /chat (typed question) ─────────────────────────

export interface ChatPromptArgs {
  systemPrompt: string;
  transcript: string;
  windowChars: number;
  chatHistory: ChatMessage[];
  historyTurns: number;
  userQuestion: string;
}

export function buildChatMessages({
  systemPrompt,
  transcript,
  windowChars,
  chatHistory,
  historyTurns,
  userQuestion,
}: ChatPromptArgs): GroqMessage[] {
  const window = tail(transcript, windowChars);
  const recent = chatHistory
    .filter((m) => m.content.trim())
    .slice(-historyTurns);
  const historyText =
    recent.length === 0
      ? "(no prior turns)"
      : recent.map((m) => `${m.role}: ${m.content}`).join("\n");
  const user = [
    "transcript_window:",
    window,
    "---",
    "chat_history:",
    historyText,
    "---",
    "user_question:",
    userQuestion,
  ].join("\n");
  return [
    { role: "system", content: systemPrompt },
    { role: "user", content: user },
  ];
}
